pluginMeasureCpu.py

```diff
@@ -75,9 +75,8 @@ def stat(self: TaskMeasureCPU, cmd: str) -> Result:
 
     def output(self, out: TftAggregateOutput) -> None:
         # Return machine-readable output to top level
-        assert isinstance(
-            self._output, PluginOutput
-        ), f"Expected variable to be of type PluginOutput, got {type(self._output)} instead."
+        if not isinstance(self._output, PluginOutput):
+            return
```

Comment:
Shouldn't we raise an exception in this case?

Reply:
The assertion (failure) already raises a kind of exception, which breaks things. So raising another kind of exception solves nothing (unless you also suggest, that the caller catches such exceptions and define exceptions to be something that can regularly happen and that callers should expect/handle -- that sounds cumbersome though).

The assertion failure happens, because the earlier run action failed to generate a result. There is really nothing to handle at this point, so ignoring is kinda fine.

The better solution will be, to handle such error cases in a cleaner way. That will happen later .But for now, this works around the crash (due to the assert *exception*).

Reply:
I guess what I am suggesting is that a crash is a reasonable thing to happen at this point if a situation occurs where generate_output failed to generate PluginOutput.

How would the evaluator know the difference between a plugin failing in this manner or a plugin not being installed in the first place if the end result of both is out.plugins does not contain the output from the run?

Reply:
I do agree the way we handle plugin results needs work in general. It looks like the success or failure of the plugin is evaluated here by checking whether stop() managed to call generate_output() or fell back on using a default BaseOutput() which seems like a pretty bad way to determine something has failed. Maybe the real fix is to address the issue here instead.

```
        if self.exec_thread.result is not None:
            r = self.exec_thread.result
            if r.returncode != 0:
                logger.error(
                    f"Error occurred while stopping {class_name}: errcode: {r.returncode} err {r.err}"
                )
            logger.debug(f"{class_name}.stop(): {r.out}")
            self._output = self.generate_output(data=r.out)
        else:
            logger.error(f"Thread {class_name} did not return a result")
            self._output = tftbase.BaseOutput("", {})
```

Reply:
The code you show already "handles" the error by logging an ERROR message. How else would you like to handle it, a hard crash towards the end of the test?

In any case, the crash currently happens later, for indirect reasons.

This patch is a workaround, to get a test "passing" at all (without a crash). I am currently heavily reworking this code, so this will be improved later.

Reply:
> How else would you like to handle it, a hard crash towards the end of the test?

In my opinion the best way to handle unexpected exceptions during the tests runs would be to write some artifact to the resulting json file that the evaluator can take in and interpret as a plugin failure.

> This patch is a workaround, to get a test "passing" at all (without a crash). I am currently heavily reworking this code, so this will be improved later.

Sounds good, I am fine with this going in for now.

Reply:
@thom311 How are you getting in a crash state. This was working for me before.

Reply:
@wizhaoredhat [here](http://file.brq.redhat.com/~thaller/ocp-tft-pr44.log) is a logfile for a testoutput. It runs commit c1714876ab148f00e30cf9ecddb4dc0b99d3e2b7. The logfile also shows the used config.yaml.

Reply:
Ah I see it is probably because that particular plugin "validateoffload" is broken

Reply:
I think we can fix validate offload. The assertion did it's job of telling us there is a bug.

Reply:
> I think we can fix validate offload. The assertion did it's job of telling us there is a bug.

A large rework is in progress, but this issue means, I cannot successfully test `main` branch and compare the run to WIP branches.

Also, the tool handles already such this failure at two distinct earlier places (right after the error happens in the plugin, and later, when the Task notices that there is no result returned). It does so by logging error messages. Granted, a better error handling shall follow, so that the failure is also recorded in the output, but at this point, an additional assertion crash is not helpful. It just breaks things, when the logging clearly indicates that there is a problem.

IMO this workaround should be merged.

Reply:
OK sounds good to me. Ok to merge.

Reply:
thank you!